src/tools/clippy/clippy_lints/src/transmute.rs

```diff
@@ -648,3 +693,57 @@ fn is_layout_incompatible<'tcx>(cx: &LateContext<'tcx>, from: Ty<'tcx>, to: Ty<'
         false
     }
 }
+
+/// Check if the the type conversion can be expressed as a pointer cast, instead of a transmute.
+fn can_be_expressed_as_pointer_cast<'tcx>(cx: &LateContext<'tcx>, e: &'tcx Expr<'_>, from_ty: Ty<'tcx>, to_ty: Ty<'tcx>) -> bool {
+    use CastKind::*;
+    matches!(
+        check_cast(cx, e, from_ty, to_ty),
+        Some(
+            PtrPtrCast
+            | PtrAddrCast
+            | AddrPtrCast
+            | ArrayPtrCast
+            | FnPtrPtrCast
+            | FnPtrAddrCast
+        )
+    )
+}
+
+/// If a cast from from_ty to to_ty is valid, returns an Ok containing the kind of the cast.
+fn check_cast<'tcx>(cx: &LateContext<'tcx>, e: &'tcx Expr<'_>, from_ty: Ty<'tcx>, to_ty: Ty<'tcx>) -> Option<CastKind> {
+    let hir_id = e.hir_id;
+    let local_def_id = hir_id.owner;
+
+    Inherited::build(cx.tcx, local_def_id).enter(|inherited| {
+        let fn_ctxt = FnCtxt::new(
+            &inherited,
+            // TODO should we try to get the correct ParamEnv?
+            ty::ParamEnv::empty(),
+            hir_id
+        );
+
+        // If we already have errors, we can't be sure we can pointer cast.
+        if fn_ctxt.errors_reported_since_creation() {
+            return None;
+        }
+
+        if let Ok(check) = CastCheck::new(
+            &fn_ctxt,
+            e,
+            from_ty,
+            to_ty,
+            // We won't show any error to the user, so we don't care what the span is here.
+            DUMMY_SP,
+            DUMMY_SP,
+        ) {
+            check.do_check(&fn_ctxt)
+                .ok()
+                // do_check's documentation says that it might return Ok and create
+                // errors in the fcx instead of returing Err in some cases.
+                .filter(|_| !fn_ctxt.errors_reported_since_creation())
```

Comment:
This seems bad... won't we start emitting errors for weird transmutes?

Reply:
My intention was that if errors happened, we would return `None`. As far as I understand, that `filter` achieves that since I negated the boolean. So if error happen in the `FnCtxt`, this code should be treating it the same as an `Err` result, unless I've missed something.

I assumed that if any errors happened in the `FnCtxt`, some other check, probably from `rustc` itself, would catch the problem and report it. Does that seem like a reasonable assumption, and does that fit with your expectations?

Reply:
My worry is that the errors have already been emitted. I haven't checked how errors could be emitted, so... I guess look for these errors and try to use the arguments on which they'd error for a cast on a transmute and see how the tests fare.

Reply:
Of all the functions called from inside `do_check`, `check_ref_cast` is the one that most clearly has a case where errors are emitted. This is because it calls [`demand_eqtype`](https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_typeck/check/demand.rs.html#72-76) which clearly can emit an error.

The following functions clearly do not emit errors:
* `CheckCast::check_ptr_ptr_cast`
* `CheckCast::check_ptr_addr_cast`
* `CheckCast::check_addr_ptr_cast`
* `CheckCast::check_fptr_ptr_cast`

I haven't confirmed whether or not the other functions called in `do_check` emit errors though.

I also had a look at the [commit the introduced `do_check` and the doc comment on it I mentioned above](https://github.com/rust-lang/rust/commit/83acebc46213e30eac3e9e71b213440104914f4c), and `check_ref_cast` was already called within `do_check` at that time, so I suspect that is what the doc comment originally referred to.

So, my next step is to add code to the tests that triggers `demand_eqtype` to emit an error, and see what happens.

Reply:
I tried adding a `dbg!` call into `demand_eqtype` so that if the error case was called, the tests would definitely fail, (due to the extra `stdout` output.) That way, we would know that the new tests would actually hit that case, so we would know whether or not we got an extra error message.

The existing tests for the new lint did not hit that case, but from looking at the code leading from `do_check` to `demand_eqtype`, it seems like a `transmute` from an array reference to a pointer to a different type should trigger the lint to  call `do_check` in a way that should trigger it. For example, `&[i32; 1]` to `*const u8`. So I added such a `transmute` to the tests, but I got an error from a different transmute lint.

While the other lint seems to prevent any issues at the moment, if that lint, or the ordering was changed in the future that could lead to a surprising bug, so it seems prudent to understand whether the `do_check` call will cause an extra error or not.

So, I temporarily placed the new lint at the top of the big `match` inside the `check_expr` implementation inside `transmute.rs`. Doing that lead to an ICE with the following error message:

```
thread 'rustc' panicked at 'assertion failed: self.find(parent).map_or(false, |n| is_body_owner(n, hir_id))', src/librustc_middle/hir/map/mod.rs:352:9
```

So without digging into what the failed assert means exactly, I would guess that something is wrong with the way that the FnCtxt was created inside clippy in this PR. But I'm not really sure. Does anyone else have more context around what that assert failing suggests?

Reply:
> but I got an error from a different transmute lint.

If that other lint is `allow`ed, would the error occur? Or is it part of this big `match`, so we never reach your new lint if the other lint is emitted? I'd be fine to just leave sufficient comments (and add a test that will fail horribly if the order is changed).

Reply:
> is it part of this big match, so we never reach your new lint if the other lint is emitted?
That is the case.

Some other tests were already failing whenever I observed the ICE, but better safe than sorry. I've added a the transmute from `&[i32; 1]` to `*const u8` case to the lint-specific test.

Reply:
Ok great! I think you can now just make this an assert instead of filtering

Reply:
Sounds good. I added a not about the panics to the comments too. All that is in [this commit](https://github.com/rust-lang/rust/pull/75098/commits/54472478fa29c2d121da469bc8f10f8b68e3e134)